istioctl/cmd/wait.go

```diff
@@ -0,0 +1,211 @@
+`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if forDelete {
+				return errors.New("wait for delete is not yet implemented")
+			} else if !forDistribution {
+				return errors.New("one of for-delete and for-distribution must be specified")
+			}
+			var versionChan chan string
+			var g *errgroup.Group
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+			if resourceVersion == "" {
+				versionChan, g = getAndWatchResource(ctx, targetResource) // setup version getter from kubernetes
+			} else {
+				versionChan = make(chan string, 1)
+				versionChan <- resourceVersion
+			}
+			// wait for all deployed versions to be contained in resourceVersions
+			t := time.NewTicker(pollInterval)
+			resourceVersions := make([]string, 1)
+			for {
+				select {
+				case newVersion := <-versionChan:
+					resourceVersions = append(resourceVersions, newVersion)
+				case <-t.C:
+					finished, err := poll(resourceVersions)
+					if err != nil {
+						return err
+					} else if finished {
+						return nil
+					}
+				case <-ctx.Done():
+					// I think this means the timeout has happened:
+					t.Stop()
+					if err := g.Wait(); err != nil {
+						return err
+					}
+					return errors.New("timeout expired before resource was distributed")
+				}
+			}
+		},
+		Args: func(cmd *cobra.Command, args []string) error {
+			if err := cobra.ExactArgs(1)(cmd, args); err != nil {
+				return err
+			}
+			targetResource = args[0]
+			return parseResource(targetResource)
+		},
+	}
+	cmd.PersistentFlags().BoolVar(&forDistribution, "for-distribution", false,
+		"wait for the designated resource to be distributed")
+	cmd.PersistentFlags().BoolVar(&forDelete, "for-delete", false,
+		"wait for the designated resource to be distributed")
+	cmd.PersistentFlags().DurationVar(&timeout, "timeout", time.Second*30,
+		"the duration to wait before failing (default 30s)")
+	cmd.PersistentFlags().Float32Var(&threshold, "threshold", 1,
+		"The ratio of distribution required for success (default 1.0)")
+	cmd.PersistentFlags().StringVar(&resourceVersion, "resource-version", "",
+		"Causes istioctl to wait for a specific version of config to become current, rather than using whatever is latest in k8s")
+	return cmd
+}
+
+func parseResource(input string) error {
+	return nil
+}
+
+func countVersions(versionCount map[string]int, configVersion string) {
+	if count, ok := versionCount[configVersion]; ok {
+		versionCount[configVersion] = count + 1
+	} else {
+		versionCount[configVersion] = 1
+	}
+}
+
+func poll(acceptedVersions []string) (bool, error) {
+	kubeClient, err := clientExecFactory(kubeconfig, configContext)
+	if err != nil {
+		return false, err
+	}
+	path := fmt.Sprintf("/debug/config_distribution?id=%s", targetResource)
+	pilotResponses, err := kubeClient.AllPilotsDiscoveryDo(istioNamespace, "GET", path, nil)
+	if err != nil {
+		return false, err
```

Comment:
It might be good to prefix this error, which will manifest if using a 1.4 _istioctl_ on a 1.3 control plane, or if the endpoint is turned off with a feature flag.